test/unittest/DatabaseQueue/DatabaseQueueTests.cpp

```diff
@@ -0,0 +1,2679 @@
+            std::make_pair(EntityId(0), EntityId(1)))));
+
+    // Precondition: The locator exists and has ID 2
+    EXPECT_CALL(database, get_entities_by_name(EntityKind::LOCATOR, dst_locator_str)).Times(2)
+            .WillRepeatedly(Return(std::vector<std::pair<EntityId, EntityId>>(1,
+            std::make_pair(EntityId(0), EntityId(2)))));
+
+    // Expectation: The insert method is called with appropriate arguments
+    InsertDataArgs args1([&](
+                const EntityId& domain_id,
+                const EntityId& entity_id,
+                const StatisticsSample& sample)
+            {
+                EXPECT_EQ(entity_id, 1);
+                EXPECT_EQ(domain_id, 0);
+                EXPECT_EQ(sample.src_ts, timestamp);
+                EXPECT_EQ(sample.kind, DataKind::RTPS_PACKETS_LOST);
+                EXPECT_EQ(dynamic_cast<const RtpsPacketsLostSample&>(sample).count, 1024);
+                EXPECT_EQ(dynamic_cast<const RtpsPacketsLostSample&>(sample).remote_locator, 2);
+            });
+
+    InsertDataArgs args2([&](
+                const EntityId& domain_id,
+                const EntityId& entity_id,
+                const StatisticsSample& sample)
+            {
+                EXPECT_EQ(entity_id, 1);
+                EXPECT_EQ(domain_id, 0);
+                EXPECT_EQ(sample.src_ts, timestamp);
+                EXPECT_EQ(sample.kind, DataKind::RTPS_BYTES_LOST);
+                EXPECT_EQ(dynamic_cast<const RtpsBytesLostSample&>(sample).count, 2048);
+                EXPECT_EQ(dynamic_cast<const RtpsBytesLostSample&>(sample).magnitude_order, 10);
+                EXPECT_EQ(dynamic_cast<const RtpsBytesLostSample&>(sample).remote_locator, 2);
+            });
+
+    EXPECT_CALL(database, insert(_, _, _)).Times(2)
+            .WillOnce(Invoke(&args1, &InsertDataArgs::insert))
+            .WillOnce(Invoke(&args2, &InsertDataArgs::insert));
+
+    // Add to the queue and wait to be processed
+    data_queue.push(timestamp, data);
+    data_queue.flush();
+}
+
+TEST_F(database_queue_tests, push_rtps_lost_no_writer)
+{
+    std::chrono::steady_clock::time_point timestamp = std::chrono::steady_clock::now();
+
+    std::array<uint8_t, 12> prefix = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12};
+    std::array<uint8_t, 4> writer_id = {0, 0, 0, 2};
+    std::array<uint8_t, 16> dst_locator_address = {16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1};
+    uint32_t dst_locator_port = 2048;
+    std::string writer_guid_str = "1.2.3.4.5.6.7.8.9.10.11.12.0.0.0.2";
+    std::string dst_locator_str = "TCPv4:[4.3.2.1]:2048";
+
+    // Build the writer GUID
+    DatabaseDataQueue::StatisticsGuidPrefix writer_prefix;
+    writer_prefix.value(prefix);
+    DatabaseDataQueue::StatisticsEntityId writer_entity_id;
+    writer_entity_id.value(writer_id);
+    DatabaseDataQueue::StatisticsGuid writer_guid;
+    writer_guid.guidPrefix(writer_prefix);
+    writer_guid.entityId(writer_entity_id);
```

Comment:
I wanted the data in the queue to be fully formed, so that the error we are detecting is really due to our expectations, and not from some other random error happening because of incomplete data, or because we modified the order or some check in the implementation.